# Assignment 3
### Due 9/21. Do four of five.

In [2]:
import urllib.request
import os

def download_data(force=False):
    """Download and extract course data from Zenodo."""
    import urllib.request
    import zipfile
    import os
    
    zip_path = 'data.zip'
    data_dir = 'data'
    
    if not os.path.exists(zip_path) or force:
        print("Downloading course data...")
        urllib.request.urlretrieve(
            'https://zenodo.org/records/16954427/files/data.zip?download=1',
            zip_path
        )
        print("Download complete")
    
    if not os.path.exists(data_dir):
        print("Extracting data files...")
        with zipfile.ZipFile(zip_path, 'r') as zip_ref:
            zip_ref.extractall(data_dir)
        print("Data extracted")
    
    return data_dir

download_data()

'data'

In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

1. 
- Open the NHANES (or Ames prices or college completion datasets, if you prefer)
- Find two categorical variables of interest (there are 198, and short descriptions are given in the `nhanes_meta_17_18.csv` file). Investigate their missing values (you don't have to focus on missing values for this analysis like we did with police use of force, but always be aware of how dirty the data are)
- Compute a contingency table for your categorical $X$ and $Y$
- Discuss any interesting patterns (or lack of one) that you observe

In [5]:
ames = pd.read_csv('./data/ames_prices.csv')
ames.head()

Order        PID  area   price  MS.SubClass MS.Zoning  Lot.Frontage  \
0      1  526301100  1656  215000           20        RL         141.0   
1      2  526350040   896  105000           20        RH          80.0   
2      3  526351010  1329  172000           20        RL          81.0   
3      4  526353030  2110  244000           20        RL          93.0   
4      5  527105010  1629  189900           60        RL          74.0   

   Lot.Area Street Alley  ... Screen.Porch Pool.Area Pool.QC  Fence  \
0     31770   Pave   NaN  ...            0         0     NaN    NaN   
1     11622   Pave   NaN  ...          120         0     NaN  MnPrv   
2     14267   Pave   NaN  ...            0         0     NaN    NaN   
3     11160   Pave   NaN  ...            0         0     NaN    NaN   
4     13830   Pave   NaN  ...            0         0     NaN  MnPrv   

  Misc.Feature Misc.Val Mo.Sold Yr.Sold Sale.Type Sale.Condition  
0          NaN        0       5    2010       WD          Normal  
1          NaN        0       6    2010       WD          Normal  
2         Gar2    12500       6    2010       WD          Normal  
3          NaN        0       4    2010       WD          Normal  
4          NaN        0       3    2010       WD          Normal  

[5 rows x 82 columns]

In [7]:
ames['Alley_na'] = ames['Alley'].isna()
ames['Alley_na'].value_counts()

Alley_na
True     2732
False     198
Name: count, dtype: int64

In [8]:
ames['MS.Zoning'].value_counts()

MS.Zoning
RL         2273
RM          462
FV          139
RH           27
C (all)      25
I (all)       2
A (agr)       2
Name: count, dtype: int64

In [11]:
pd.crosstab(ames['Alley_na'], ames['MS.Zoning'])

MS.Zoning  A (agr)  C (all)  FV  I (all)  RH    RL   RM
Alley_na                                               
False            0        3  52        0   4    40   99
True             2       22  87        2  23  2233  363

In [13]:
pd.crosstab(ames['Alley_na'], ames['MS.Zoning'], normalize='columns')

MS.Zoning  A (agr)  C (all)        FV  I (all)        RH        RL        RM
Alley_na                                                                    
False          0.0     0.12  0.374101      0.0  0.148148  0.017598  0.214286
True           1.0     0.88  0.625899      1.0  0.851852  0.982402  0.785714

In [15]:
ames.loc[:,['Alley_na','MS.Zoning']].groupby('MS.Zoning').mean()

Alley_na
MS.Zoning          
A (agr)    1.000000
C (all)    0.880000
FV         0.625899
I (all)    1.000000
RH         0.851852
RL         0.982402
RM         0.785714

Missing values dominate: Almost all houses 93% have no recorded alley access Missing. Only 7% report an alley type.

In RL Residential Low Density, the largest category, 97% of houses have missing alley info, showing that alleys are rare in low-density zones.

In RM Residential Medium Density, alleys are more common 99 non-missing vs 363 missing, 21% alleys present.

In FV Floating Village Residential, alleys are also relatively more common 37% alleys present.

For C all Commercial, about 12% 3 out of 25 have alley info.

Alley presence seems to vary somewhat by zoning. In higher-density or special-use zones RM,FV, and alleys are more likely to be present compared to the dominant low-density RL zone.

2. 
- Open the NHANES dataset
- Find a categorical and numeric variable of interest (there are 198, and short descriptions are given in the `nhanes_meta_17_18.csv` file). Investigate their missing values (you don't have to focus on missing values for this analysis, but always be aware of them)
- Make descriptive tables and grouped kernel density plots to represent the variation in your numeric $Y$ conditional on your categorical $X$
- Discuss any interesting patterns (or lack of one) that you observe

In [16]:
meta = pd.read_csv("./data/nhanes_meta_17_18.csv")
meta.head()

VariableNameLong Variable     Type  Format  Informat  \
0          GeneralHealthCondition   HSD010  Numeric     NaN       NaN   
1    EverBreastfedOrFedBreastmilk   DBQ010  Numeric     NaN       NaN   
2     AgeStoppedBreastfeedingdays   DBD030  Numeric     NaN       NaN   
3          AgeFirstFedFormuladays   DBD041  Numeric     NaN       NaN   
4  AgeStoppedReceivingFormuladays   DBD050  Numeric     NaN       NaN   

                                 Label Source VariableName  \
0             General health condition    HSQ       HSD010   
1     Ever breastfed or fed breastmilk    DBQ       DBQ010   
2      Age stopped breastfeeding(days)    DBQ       DBD030   
3          Age first fed formula(days)    DBQ       DBD041   
4  Age stopped receiving formula(days)    DBQ       DBD050   

                              SASLabel  \
0             General health condition   
1     Ever breastfed or fed breastmilk   
2      Age stopped breastfeeding(days)   
3          Age first fed formula(days)   
4  Age stopped receiving formula(days)   

                                         EnglishText  \
0  Next I have some general questions about {your...   
1  Now I'm going to ask you some general question...   
2  How old was {SP} when {he/she} completely stop...   
3  How old was {SP} when {he/she} was first fed f...   
4  How old was {SP} when {he/she} completely stop...   

                                              Target  \
0   Both males and females 12 YEARS -\n\n\t\t\t15...   
1   Both males and females 0 YEARS -\n\n\t\t\t6 Y...   
2   Both males and females 0 YEARS -\n\n\t\t\t6 Y...   
3   Both males and females 0 YEARS -\n\n\t\t\t6 Y...   
4   Both males and females 0 YEARS -\n\n\t\t\t6 Y...   

     VariableNameLong_variable_df Source_variable_df EnglishInstructions  \
0          GeneralHealthCondition                HSQ                 NaN   
1    EverBreastfedOrFedBreastmilk                DBQ                 NaN   
2     AgeStoppedBreastfeedingdays                DBQ                 NaN   
3          AgeFirstFedFormuladays                DBQ                 NaN   
4  AgeStoppedReceivingFormuladays                DBQ                 NaN   

  HardEdits  Recoded CustomRecoding  nNonNA  
0       NaN     True            NaN  5964.0  
1       NaN     True          YesNo  1069.0  
2       NaN    False            NaN   862.0  
3       NaN    False            NaN  1077.0  
4       NaN    False            NaN   891.0

3. 
We showed that the mean and median could be discovered by minimizing various kinds of loss functions; this is what machine learning is. To make a prediction $\hat{y}(z)$ of $Y$ when $X=z$, minimize the mean squared error:
$$
MSE(\hat{y}(z)) = \dfrac{1}{N} \sum_{i=1}^N \left\lbrace y_i - \hat{y}(z) \right\rbrace^2 \frac{1}{h}k\left(\frac{z-x_i}{h}\right)
$$
Show that the solution to this problem is the LCLS/Naradaya-Watson estimator.

4. 
- Write a class or set of functions that implement the LCLS/Naradaya-Watson estimator, using the Silverman plug-in estimate for the conditioning variable $X$ as the bandwidth.
- From one of the course data sets, find two numeric variables of interest, analyze their relationship with the the LCLS/Naradaya-Watson estimator, and discuss your results.

5. 
- In any of the available data sets, investigate the relationships between pairs of variables $(X,Y)$ with a scatterplot and CEF (for example, price on area)
- Is this relationship plausibly causal, or are there missing variables that might explain at least part of the relationship between your variables? These can be "conceptual" rather than "practical"; for example, 'talent' or 'grit' probably explain education outcomes, but are almost impossible to measure. We are asking whether there are hypothetical **threats to causal identification** of the effect of $X$ on $Y$.
- Explain how, regardless of the threat to causal identification, you can still use your model to predict $Y$ given $X$, as long as you don't intervene in the system to control the outcome